### ISSUE TITLE
Running same axe-storybook command produces inconsistent results

### ISSUE THREAD

**Issue by tstewart15** [chanzuckerberg/axe-storybook-testing#34]
I'm still investigating this issue, so apologies if my description is vague.

What's currently happening is I'm running the same axe-storybook command on an unchanged Storybook directory and getting inconsistent results.

My command:
```
yarn axe-storybook --build-dir ./axe-storybook --failing-impact moderate --pattern TabButton
```
(the `--pattern` is likely insignificant; I've been noticing these inconsistent results for a while and applied this pattern just to try and narrow this down)

Sometimes, this passes:
![image](https://user-images.githubusercontent.com/8942170/133516187-3b3c3521-785d-4e7f-887f-f135b60c3f6c.png)

Other times, it reports issues:
![image](https://user-images.githubusercontent.com/8942170/133516085-dcf524e0-436d-4ed6-b328-632d1522768b.png)

The story file I'm testing:
```js
import { getRenderOptions } from '@linkedin/storybook-glimmer';
import {TabButton} from '../test/glimmer/index';
import defaultData from '/Users/tomstewa/Multiproducts/3/guest-ui-lib_trunk/test/data/components/common/tabs/tab-button/default.json';


export default {
  title: 'common/tabs/TabButton',
  parameters: {
    SetupPreview: { storyComponent: TabButton },
    axe: {
      disabledRules: ['heading-order', 'valid-lang'],
    }
  }
};


export const TabButton_with_dataFile_default = (args) => {
  return {
    componentClass: TabButton,
    renderOptions: getRenderOptions(args),
    assets: {
      css: 'css/components/common/tabs/tab-button-test.css',
      js: ''
    }
  };
};
TabButton_with_dataFile_default.args = defaultData;

```

I've tried to test this by setting `--headless false` and using different browsers; the inconsistent results still happen:
```
yarn axe-storybook --build-dir ./axe-storybook --failing-impact moderate --pattern TabButton --headless false --browser firefox
```

Not sure if anyone will be able to replicate this issue on their end, so any advice on how I can debug this would be appreciated. I'm currently trying to figure out if/how I can set a debugger breakpoint in the playwright-generated browser.

**Comment by ahuth** [comment]
Thanks for the report, and this is super weird.

A couple ideas:
- What version of axe-storybook-testing are you on? I recently released versions 4.0.0 and 4.0.1. Not sure if this is a regression or an existing bug.
- If you run Storybook and with its [Accessibility Addon](https://storybook.js.org/addons/@storybook/addon-a11y), does it find any Axe violations? (In general running Storybook manually should be pretty close to stopping playwright in the middle of a run)

**Comment by tstewart15** [comment]
Great points, thanks @ahuth 
- I am on version 3.0.2. Let me try updating to 4.0.1 (are there any major version migrations necessary?)
- The correct axe response in this example is to have the 2 violations, as seen here: 
![image](https://user-images.githubusercontent.com/8942170/133681878-47041fc0-48fd-4712-83a7-717eecdeb7fa.png)
Which means this bug has to do with axe-storybook-testing erroneously missing the violations in the component.
- There could be delays happening since I believe these components are client-side rendered. What's worrisome is the inconsistency / race condition that could be happening. Will try to investigate this further after bumping to 4.0.1.

**Comment by ahuth** [comment]
No migrations are necessary to upgrade to 4.0.2. Main change was dropping Node 10 support. There were also some internal changes, but I don't **_think_** they could cause this.

If the problem is a rendering delay, we could add the option for to wait some configured time before running axe.

We haven't seen any issues like this with React components, though 🤔 (maybe React blocks the browser thread, but whatever library is used here does not?).

For us the problem is almost always animations, which can be solved by making them take the [prefers-reduced-motion media query](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion) into account.

**Comment by tstewart15** [comment]
This rendering delay issue is somewhat confirmed by the fact that if I load the Storybook page with the "Accessibility" tab selected by default, then the issues are missed:
![image](https://user-images.githubusercontent.com/8942170/133837819-d6e5f25f-4015-47fa-b763-7a42e5722872.png)

But if I switch to the Accessibility tab after the component content is loaded, then the issues appear:
![image](https://user-images.githubusercontent.com/8942170/133837894-7effc20c-d340-46ab-ba6e-94e2e5a069d9.png)

**Comment by tstewart15** [comment]
Also, confirmed issue is still occurring after 4.0.1 update.

**Comment by ahuth** [comment]
Thanks for looking into it.

Do you think a per-story option to configure a delay after rendering and before running axe would work well for your use case?

Waiting an arbitrary amount of time isn't ideal. But I can't think of another way to solve this.

**Comment by ahuth** [comment]
Actually, maybe an alternative (and better) solution is to add an option to wait for arbitrary selectors

```js
// SomeComponent.stories.jsx

export const parameters = {
  axe: {
    waitForSelector: 'button[role="tab"]',
  },
};
```

**Comment by tstewart15** [comment]
Yes, I like this idea a lot! We may also want to add some config for the "timeout" (time before "waitForSelector" fails, described at https://playwright.dev/docs/api/class-page/#page-wait-for-selector)

**Comment by ahuth** [comment]
Implemented in #35.

@tstewart15, if you have time, can you test out the alpha version [v4.1.0-alpha.1](https://www.npmjs.com/package/@chanzuckerberg/axe-storybook-testing/v/4.1.0-alpha.1)? If you don't have time, that's fine and I'll publish the full version later today.

**Comment by ahuth** [comment]
Regarding config for the `waitForSelector` timeout, I think we're okay without it because there's already a global config for axe-storybook-testing for `--timeout`.

I tested out what happens when waitForSelector takes longer than that timeout, and it still fails as expected.

**Comment by tstewart15** [comment]
@ahuth I won't be able to pull your changes into my current project due to the 3rd party package approval process for my company. I could pull into a local project, but it won't verify much compared to your demo testing. I will wait for 4.1.0 version to get approval.

**Comment by ahuth** [comment]
Published [v4.1.0](https://github.com/chanzuckerberg/axe-storybook-testing/releases/tag/v4.1.0).

Let me know if this solves the issue or not!

**Comment by tstewart15** [comment]
Hey @ahuth the `waitForSelector` solution is working well, except for our tests which contain hidden/invisible elements. This is due to the default `state` option for [page.waitForSelector](https://playwright.dev/docs/api/class-page/#page-wait-for-selector) being "visible". Can we either hard-code this option to "attached" for our use case (since a11y / screen-reader testing generally involves elements being in the DOM but not necessarily visible) or expose an additional option on axe-storybook-testing to configure this `state` setting?

**Comment by ahuth** [comment]
Thanks for the info. I'll change it to `attached`. Can't think of a reason to not allow waiting for hidden elements for this library.

**Comment by ahuth** [comment]
Actually, just realized that if I expose the `state` option, people could then also wait for an element to not be in the DOM (which came up in https://github.com/chanzuckerberg/axe-storybook-testing/pull/35#issuecomment-923890137

**Comment by tstewart15** [comment]
Thanks @ahuth! I'll pull this in when 4.2.0 is published

**Comment by ahuth** [comment]
Just published v4.1.1 which includes this @tstewart15 

(I counted configuring for `attached` as a bugfix, instead of a new feature)
